clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -72,16 +74,28 @@ public void prepare() throws IOException {
             transportLayer.handshake();
         if (transportLayer.ready() && !authenticator.complete())
             authenticator.authenticate();
+        if (transportLayer.ready() && authenticator.complete())
+            state = ChannelState.READY;
     }
 
     public void disconnect() {
         disconnected = true;
         transportLayer.disconnect();
     }
 
+    public void state(ChannelState state) {
+        this.state = state;
+    }
+
+    public ChannelState state() {
+        return this.state;
+    }
 
     public boolean finishConnect() throws IOException {
-        return transportLayer.finishConnect();
+        boolean connected = transportLayer.finishConnect();
+        if (connected)
+            state = ready() ? ChannelState.READY : ChannelState.AUTHENTICATE;
```

Comment:
Seems like we transition to AUTHENTICATE before the TLS handshake (if TLS is enabled). If TLS authentication fails, will we also get a warning then? I think that would be good, but just checking as previous proposals didn't cover TLS.